Convert all content types to string

ibhagwan opened the pull request:
`fzf_exec` now converts all content types (tables/funcs) to shell command strings this greatly simplifies the code as all picker inputs are now shell commands which can be used in fzf's reload, execute-silent, etc.

This will also enable us to combine pickers easily with semicolon or logical and, e.g. `cmd1; cmd2` or `cmd1 && cmd2`.

Fixes #2058 and prepares the ground for #1879, #2058.

ibhagwan commented:
> `:FzfLua live_grep rg_glob=true multiprocess=false` glob don't work after [624ae3b](https://github.com/ibhagwan/fzf-lua/commit/624ae3bf06283a1151fc5aaa7f603de6f1f8f35d).

Ty @phanen, I’ll take a look, still a lot more  work but I’m happy with the code cleanup.

ibhagwan commented:
> live_grep_st use a function rather a pre-built string:

Correct, but I have to find a generic solution for this that isn’t a hack or maybe a condition in live_grep which changes `fn_reload` based on `opts.multiprocess`?

`mutliprocess=false` is just for testing/debugging issues anyways.

phanen commented:
I remember I've been using multiprocess=false because I don't need file_icon/git_icon and want glob feature.


```diff
diff --git a/lua/fzf-lua/make_entry.lua b/lua/fzf-lua/make_entry.lua
index 41e701d..7f30605 100644
--- a/lua/fzf-lua/make_entry.lua
+++ b/lua/fzf-lua/make_entry.lua
@@ -230,7 +230,7 @@ M.preprocess = function(opts)
 
   -- nifty hack to avoid having to double escape quotations
   -- see my comment inside 'live_grep' initial_command code
-  if opts.argv_expr then
+  if opts.argv_expr and opts.cmd then
     opts.cmd = opts.cmd:gsub("{argv.*}",
       function(x)
         local idx = x:match("{argv(.*)}")
diff --git a/lua/fzf-lua/providers/grep.lua b/lua/fzf-lua/providers/grep.lua
index c667d7b..537d459 100644
--- a/lua/fzf-lua/providers/grep.lua
+++ b/lua/fzf-lua/providers/grep.lua
@@ -295,18 +295,23 @@ M.live_grep = function(opts)
   end
 
   -- signal to preprocess we are looking to replace {argvz}
-  opts.argv_expr = true
+  opts.argv_expr = opts.multiprocess
 
   -- this will be replaced by the appropriate fzf
   -- FIELD INDEX EXPRESSION by 'fzf_exec'
-  opts.cmd = get_grep_cmd(opts, core.fzf_query_placeholder, 2)
-  if not opts.cmd then return end
+  opts.cmd = opts.multiprocess and
+      get_grep_cmd(opts, opts.multiprocess and core.fzf_query_placeholder or nil, 2)
 
   -- search query in header line
   opts = core.set_title_flags(opts, { "cmd", "live" })
   opts = core.set_header(opts, opts.headers or { "actions", "cwd" })
   opts = core.set_fzf_field_index(opts)
-  core.fzf_live(opts.cmd, opts)
+  core.fzf_live(opts.cmd or function(s)
+    -- can be nil when called as fzf initial command
+    local query = s[1] or ""
+    opts.no_esc = nil
+    return get_grep_cmd(opts, query, true)
+  end, opts)
 end
 
 M.live_grep_glob = function(opts)
```

ibhagwan commented:
> I remember I've been using multiprocess=false because I don't need file_icon/git_icon and want glob feature while live_grep_native don't support glob.

You’re always at a performance disadvantage this way, why not use multiprocess=true with git_icons=false and file_icons=false?

phanen commented:
I don't know... `FzfLua live_grep rg_glob=true multiprocess=false` is significantly faster than `FzfLua live_grep rg_glob=true multiprocess=true`

ibhagwan commented:
> I don't know... `FzfLua live_grep rg_glob=true multiprocess=false` is significantly faster than `FzfLua live_grep rg_glob=true multiprocess=true`

Weird, it shouldn’t be, multiprocess should always be faster as it’s an external independent command.

ibhagwan commented:
> live workspace symbol also broken.

Ty for the diff fixes, I’ll add these later today.

phanen commented:
Damn, I forgot to disable `temp-throttle.service`. So my words may be wrong, I will test it again.

ibhagwan commented:
Pushed your fixes in https://github.com/ibhagwan/fzf-lua/commit/497ab31d6d1083e9d80589ae4441b6533adf9750, ty @phanen!

ibhagwan commented:
@phanen, fixed some more stuff, still have an issue with multiprocess=false and file_icons=mini which I’m able to solve with vim.schedule by uncommenting this line:
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/libuv.lua#L216

But that presents another issue, it seems that when I use vim.scheudule there the fzf pipe closes before all the reads are complete (maybe fzf closes it with process exit?), if you run `:FzfLua live_grep multiprocess=false` and type quickly you will sometimes encounter partial/empty results, or if you run `:FzfLua files` on a large folder the last read will get cut resulting in less number of files.

Had to go out so I left it commented for now but once I solve that it would also remove the need for the `ft_match` mock as calls to `fn_transform` will be executed outside of  the “fast event”.

Once I solve this we can “retire” both of these functions (I think):
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/path.lua#L612
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/path.lua#L639

ibhagwan commented:
Seems the issue is the pipe gets closed before finalizing all writes so this line is hit:
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/shell.lua#L328

Probably some race condition causing this to trigger:
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/shell.lua#L315-L320

ibhagwan commented:
Fixed in https://github.com/ibhagwan/fzf-lua/pull/2152/commits/f37475181b62da1f434ac1cdb31653d77b66c75a, finish would be called before any read were scheduled so `write_cb_count` would be 0, needed to add a `read_cb_count`.

ibhagwan commented:
Only issue left `vim.filetype.match` is failing the tests (commented for now):
https://github.com/ibhagwan/fzf-lua/blob/dd10409aadce6f3f2f4c8e399d5d86f315e45acc/lua/fzf-lua/devicons.lua#L518-L521

Obviously we can leave it as is but I’d like to find out what’s happening here as this may indicate other issues when users calling vim functions with fzf-lua’s API.

phanen commented:
Some ideas about improvement:
* `rg_glob=true` don't require `fn_transform`, only need `fn_preprocess` to just parse the glob:
https://github.com/ibhagwan/fzf-lua/blob/01ef2bbda511c6e6c02a3c6b39c5b6fdd8e7ea7b/lua/fzf-lua/make_entry.lua#L216-L218
Although st and mt use a different codepath here:
https://github.com/ibhagwan/fzf-lua/blob/4350a64466701d71016957d2a39aec16b2ac8170/lua/fzf-lua/providers/grep.lua#L64-L75
* Not necessary to do `fn_preprocess` then `uv.spawn`, there's a option to `change:transfrom:nvim -l rpc.lua {q}` to parse the glob only then generate a `stringify_cmd` to be executed by fzf-lua directly. With this approach we can make glob parse fast as `live_grep_native`.

ibhagwan commented:
I also feel the logic of my last commit https://github.com/ibhagwan/fzf-lua/pull/2152/commits/4350a64466701d71016957d2a39aec16b2ac8170  should b changed: 
- multiprocess = true should always attempt multiprocess (since API users can also use it now)
- multiprocess = 1 means optional processing with make_entry.file/preprocess
- the logic of this should be moved to normalize_opts

This way we will restore the old behavior of multiprocess=false without icons/formatting

ibhagwan commented:
@phanen, https://github.com/ibhagwan/fzf-lua/pull/2152/commits/8aead20f5ba7e27e4e900c3eb94aff27a89a2380 restores the speed to:
```lua
-- debug=1 will now also print the command at the first line, same as multiprocess+debug
:FzfLua files profile=default multiprocess=false file_icons=false debug=1
```

ibhagwan commented:
> One way is just not prepending icons (seems what snacks/telescope did).

I haven’t looked at Telescode’s code in forever but when it started it prepended icons, extmark was only added to neovim much later.


> This example append icons because prepend icons will break fzf match highlighting. I'm not sure if it's solvable.

Very nice, that will definitely simplify the code and performance but we’ll have to also find a solution for formatters, path_shorten, etc.

Path shorten for example could get very complicated as the entire line isn’t always visible in fzf - it could display …<rest of line>

ibhagwan commented:
https://github.com/ibhagwan/fzf-lua/pull/2152/commits/6af4ff02c720a17038459259466032645daf6254

@phanen, I updated libuv.spawn again, added queue borrowed from snacks.nvim, improves a bit but still feels laggy with bulk updates when using `multiprocess=false`, queue is enabled by default try both and see the difference:
```lua
-- Long lag but super fast, all entries appear at once
:FzfLua files multiprocess=false use_queue=false
-- Using queue, more responsive but still laggy with "bulk" updates
:FzfLua files multiprocess=false
-- Multi process (implied true), zero lag, perhaps a bit slower?
:FzfLua files
```

ibhagwan commented:
@phanen, will you try out my latest commit and lmk what you think? I added a few options but I'm still mighty confused why it's not freeing the lag fully with `multiprocess=false`, even when using a queue and doing the writes serially with a coroutine if I uncomment one of the sleeps below the UI will freeze until it's over, not sure why that is, why does fzf UI refuse to update when the CPU is idle?


```lua
-- uses queue and throttling by default
:FzfLua files multiprocess=false use_queue=nil throttle=nil
-- queue without throttle, not sure if the responsiveness is better
:FzfLua files multiprocess=false use_queue=nil throttle=false
-- no queue with throttle 
:FzfLua files multiprocess=false use_queue=false throttle=nil
-- no queue, no throttle, fastest but freezes the UI until all results are in
:FzfLua files multiprocess=false use_queue=false throttle=false
```

Try uncommenting either of these lines, and you'll see what I mean:
https://github.com/ibhagwan/fzf-lua/blob/c3c44d477e13e24f6aa51693d3a842a13cbe9bfa/lua/fzf-lua/shell.lua#L355-L356
https://github.com/ibhagwan/fzf-lua/blob/c3c44d477e13e24f6aa51693d3a842a13cbe9bfa/lua/fzf-lua/shell.lua#L358-L361

I tried queueing the `uv.write` calls in threads using  `uv.new_work|uv.work_queue` and still nada, do you have any idea?

Overall I find the performance of `multiprocess=false` acceptable now (seems much better) but I would prefer to solve this mystery.

ibhagwan commented:
Just had a shower thought, I’ve been barking up the wrong tree :)

The issue comes from the spawning process code  ( `nvim -l` ) code, I’ll try to thread/coroutinify the below:
https://github.com/ibhagwan/fzf-lua/blob/7e700679918f173d698c60035e48ea5d7a2fb6a4/lua/fzf-lua/rpc.lua#L40-L53

ibhagwan commented:
This is even more strange, I tried running the  neovim proxy `uv.listen` in a thread, didn't work for some reason so I made `rpc_nvim_exec_lua` run in it's own process:
1. The main instance
2. The shell command instance (nvim -l) acting as proxy between fzf and the main instance (pipe)
3. Another process executed by (2) using `jobstart` which does a single `vim.rpcrequest`  and still the same behavior

This makes me think having `vim.rcprequest` called in the main instance is the issue, tried to chase a lead with using pure lua msgpack-rpc instaed of `vim.rpcrequest` but I'm not sure if that would help.

ibhagwan commented:
Alright, after a few days of messing with this I’m giving up on the idea to make multiprocess=false a lag free experience as it seems the issue stems from neovim being single threaded, so without using threads (which have no neovim runtime) it’s never going to be as smooth as multiprocess=true (which most pickers should use when processing shell output anyway).

Back to the original mission, 2 tasks left.

ibhagwan commented:
> there's a bug(maybe blame to my commit). loop: `ctrl-c` -> `FzfLua live_grep` -> `ctrl-c` -> `FzfLua resume`. no result, sometime error

I’ll check, I have yet to add testing for fzf_live, once we figure this out we should a test case for this too.

ibhagwan commented:
@phanen, affer enough ctrl-c <-> resume I’m getting this asset:
https://github.com/ibhagwan/fzf-lua/blob/32e19e0896458450898bcc910a4b9ab23323b844/lua/fzf-lua/shell.lua#L34

Obviously with resume, clear protected is never called but seems that new functions are registered each call until the ring buffer is full, I need to find a new solutions for this, I don’t like the “protected” idea but I also don’t want a really large buffer limit occupying a lot of memory, need to thing of something else.

ibhagwan commented:
Not fully fixed, need to check why functions are re-registered, but fixed the resume, deprecated fzf < 0.35, using `start` event instead of modifying the initial query, that's what I used for my shell `ctrl-f` shortcut too, it's much more robust:
https://github.com/ibhagwan/fzf-lua/blob/c8b53610c868ead5b8c4057711d3e79ddf385a17/lua/fzf-lua/core.lua#L1230-L1240

ibhagwan commented:
The issue was that all calls to stringify were protected indiscriminately, my last commit is still bugged and will also bug out on skim as there's no reload event so the actions may get overriden for example with `git_status`, I need to think of a better solution for storing and clearing callback function pointers.

phanen commented:
After cc480bf3adc38c0af827d752d0b2bbcc8d36fbb8, `actions.start` cannot be used in custom picker.
e.g. https://github.com/phanen/fzf-lua-overlay/blob/77c6bc0d6876e5fa75888ec9d326b4743655f32d/lua/fzf-lua-extra/providers/runtime.lua#L17-L21

(Btw we can have an option for fzf_live's fzf_field_index, https://github.com/ibhagwan/fzf-lua/blob/c7780cb18e438aa27a3f48d40559bc6f4f8f3e0d/lua/fzf-lua/core.lua#L188)

ibhagwan commented:
> After [cc480bf](https://github.com/ibhagwan/fzf-lua/commit/cc480bf3adc38c0af827d752d0b2bbcc8d36fbb8), actions.start cannot be used in custom picker. e.g. https://github.com/phanen/fzf-lua-overlay/blob/77c6bc0d6876e5fa75888ec9d326b4743655f32d/lua/fzf-lua-extra/providers/runtime.lua#L17-L21

I wonder why it doesn’t work, I’ll have to check, the call to convert should work with a custom picker too:

> fzf_live sends true in convert_acrions
https://github.com/ibhagwan/fzf-lua/blob/c7780cb18e438aa27a3f48d40559bc6f4f8f3e0d/lua/fzf-lua/core.lua#L192

> Which calls convert exec silent:
https://github.com/ibhagwan/fzf-lua/blob/c7780cb18e438aa27a3f48d40559bc6f4f8f3e0d/lua/fzf-lua/core.lua#L215-L218

ibhagwan commented:
Im too deprived of sleep lol, just realized you’re using fzf_live which now makes use of the start event too:
https://github.com/ibhagwan/fzf-lua/blob/c7780cb18e438aa27a3f48d40559bc6f4f8f3e0d/lua/fzf-lua/core.lua#L1233-L1234

Now this comment makes sense (which is the right solution):
> (Btw we can have an option for fzf_live's fzf_field_index,

Another (breaking) change is that now the callback param is an array (I updated the wiki), so the callbacks should be:
```lua
require 'fzf-lua'.fzf_live(
  function(args)
    local q = args[1]
   …
```

ibhagwan commented:
> This work for me
> 
> ```diff
> diff --git a/lua/fzf-lua/core.lua b/lua/fzf-lua/core.lua
> index f718696..97a6d84 100644
> --- a/lua/fzf-lua/core.lua
> +++ b/lua/fzf-lua/core.lua
> @@ -1231,7 +1231,7 @@ M.setup_fzf_interactive_flags = function(command, fzf_field_index, opts)
>        string.format("change:reload:%s%s", no_query_condi, reload_command)))
>      if utils.has(opts, "fzf", { 0, 35 }) then
>        opts._fzf_cli_args = opts._fzf_cli_args .. string.format(" --bind=%s",
> -        libuv.shellescape(string.format("start:reload:%s%s", no_query_condi, reload_command)))
> +        libuv.shellescape(string.format("start:+reload:%s%s", no_query_condi, reload_command)))
>      end
>    end
>  
> ```

This is a great fix! we should implement this in the convert actions too:

https://github.com/ibhagwan/fzf-lua/blob/f972ad787ee8d3646d30000a0652e9b168a90840/lua/fzf-lua/core.lua#L1255-L1269



The only thing I can’t recall is if this `start:+…` logic new in fzf?

Do you recall which fzf version added this logic so we can limit it (if it requires > 0.36)?

ibhagwan commented:
Apparently the + logic in binds existed since fzf 0.20, I’ll make the change.

ibhagwan commented:
Should also change `line_query` and hide profile events to fzf_cli_args with + so we don’t override user events:
https://github.com/ibhagwan/fzf-lua/blob/3d73bf5ce9cc3dc0e358534a6dd4ec80f77d2387/lua/fzf-lua/config.lua#L924

phanen commented:
Even more..
```diff
diff --git a/lua/fzf-lua/config.lua b/lua/fzf-lua/config.lua
index 1225b88..74bb403 100644
--- a/lua/fzf-lua/config.lua
+++ b/lua/fzf-lua/config.lua
@@ -922,7 +922,7 @@ function M.normalize_opts(opts, globals, __resume_key)
   elseif opts.line_query then
     utils.map_set(opts, "winopts.preview.winopts.cursorline", true)
     utils.map_set(opts, "keymap.fzf.change",
-      "transform:" .. FzfLua.shell.stringify_data(function(q, _, _)
+      "transform:+" .. FzfLua.shell.stringify_data(function(q, _, _)
         local lnum = q[1]:match(":(%d+)$")
         local new_q, subs = q[1]:gsub(":%d*$", "")
         -- No subs made, no ":" at end of string, do nothing
diff --git a/lua/fzf-lua/previewer/builtin.lua b/lua/fzf-lua/previewer/builtin.lua
index 2641019..c8d97e7 100644
--- a/lua/fzf-lua/previewer/builtin.lua
+++ b/lua/fzf-lua/previewer/builtin.lua
@@ -469,7 +469,7 @@ function Previewer.base:zero(_)
   -- https://github.com/junegunn/fzf/issues/3516
   --
   self._zero_lock = self._zero_lock or path.normalize(vim.fn.tempname())
-  local act = string.format("execute-silent(mkdir %s && %s)",
+  local act = string.format("+execute-silent(mkdir %s && %s)",
     libuv.shellescape(self._zero_lock),
     shell.stringify_data(function(_, _, _)
       vim.defer_fn(function()
diff --git a/lua/fzf-lua/profiles/hide.lua b/lua/fzf-lua/profiles/hide.lua
index 635ddcc..fcd5068 100644
--- a/lua/fzf-lua/profiles/hide.lua
+++ b/lua/fzf-lua/profiles/hide.lua
@@ -84,7 +84,7 @@ return {
         return act
       end, opts.actions)
       -- Hijack the resize event to reload buffer/tab list on unhide
-      opts.keymap.fzf.resize = "transform:" .. shell.stringify_data(function(_, _, _)
+      opts.keymap.fzf.resize = "transform:+" .. shell.stringify_data(function(_, _, _)
         if opts._unhide_called then
           opts._unhide_called = nil
           if opts.__reload_cmd then
diff --git a/lua/fzf-lua/providers/buffers.lua b/lua/fzf-lua/providers/buffers.lua
index 4017991..733621d 100644
--- a/lua/fzf-lua/providers/buffers.lua
+++ b/lua/fzf-lua/providers/buffers.lua
@@ -384,7 +384,7 @@ M.tabs = function(opts)
   if opts.locate and utils.has(opts, "fzf", { 0, 36 }) then
     -- Set cursor to current buffer
     utils.map_set(opts, "keymap.fzf.load",
-      "transform:" .. FzfLua.shell.stringify_data(function(_, _, _)
+      "transform:+" .. FzfLua.shell.stringify_data(function(_, _, _)
         local pos = 0
         for tabnr, tabh in ipairs(vim.api.nvim_list_tabpages()) do
           pos = pos + 1
```

ibhagwan commented:
> Even more..

It is better to also move these to fzf_cli_args to nor override the users keymap.fzf.

ibhagwan commented:
@phanen, almost there, to fix the gc collection I had to change to opening files without a previewer, for some odd reason when using the builtin previewer the gc isn't called?
https://github.com/ibhagwan/fzf-lua/blob/2f87829214e577728dde85ecc7384c8bb54ac219/tests/win_spec.lua#L49

The api event test is also failing on Windows, unsure why will have to test further, tricky as I can only test via the CI as I don't have access to a Windows VM anymore...

phanen commented:
I believe this related to the side effect of something:
```diff
diff --git a/lua/fzf-lua/core.lua b/lua/fzf-lua/core.lua
index 6e05507..defb29b 100644
--- a/lua/fzf-lua/core.lua
+++ b/lua/fzf-lua/core.lua
@@ -592,6 +592,9 @@ M.create_fzf_colors = function(opts)
 end
 
 M.create_fzf_binds = function(opts)
+  do
+    local _ = "execute-silent:" .. shell.stringify_data(function() end, {})
+  end
   local binds = opts.keymap.fzf
   if not binds or utils.tbl_isempty(binds) then return {} end
   local combine, separate = {}, {}
diff --git a/lua/fzf-lua/test/helpers.lua b/lua/fzf-lua/test/helpers.lua
index a6c8478..5c4ccfa 100644
--- a/lua/fzf-lua/test/helpers.lua
+++ b/lua/fzf-lua/test/helpers.lua
@@ -132,7 +132,10 @@ M.new_child_neovim = function()
       require("fzf-lua").setup(vim.tbl_deep_extend("keep", ..., {
         %s
         winopts = {
-          on_create = function() _G._fzf_lua_on_create = true end,
+          on_create = function()
+            _G._fzf_lua_on_create = true
+            vim.defer_fn(function() _G._fzf_load_called = true end, 20)
+          end,
           on_close = function()
             _G._fzf_lua_on_create = nil
             _G._fzf_postprocess_called = nil
@@ -141,10 +144,10 @@ M.new_child_neovim = function()
         },
         -- TODO: why is gc called with builtin with this one?
         -- keymap = { fzf = { true, load = function() _G._fzf_load_called = true end } }
-        fzf_cli_args = "--bind=" .. FzfLua.libuv.shellescape("load:+execute-silent:"
-          .. FzfLua.shell.stringify_data(function(_, _, _)
-              _G._fzf_load_called = true
-            end, {}))
+        -- fzf_cli_args = "--bind=" .. FzfLua.libuv.shellescape("load:+execute-silent:"
+        --   .. FzfLua.shell.stringify_data(function(_, _, _)
+        --       _G._fzf_load_called = true
+        --     end, {}))
       }))
     ]])
         -- using "FZF_DEFAULT_OPTS" hangs the command on the
diff --git a/tests/win_spec.lua b/tests/win_spec.lua
index 322dea3..b193174 100644
--- a/tests/win_spec.lua
+++ b/tests/win_spec.lua
@@ -46,8 +46,8 @@ T["win"]["hide"]["ensure gc called after win hidden (#1782)"] = function()
       end
     end
     -- TODO: why isn't gc called with the builtin previewr
-    -- child.lua([[FzfLua.files{ previewer = 'builtin' }]])
-    child.lua([[FzfLua.files{ previewer = false }]])
+    child.lua([[FzfLua.files{ previewer = 'builtin' }]])
+    -- child.lua([[FzfLua.files{ previewer = false }]])
     child.wait_until(function()
       return child.lua_get([[_G._fzf_load_called]]) == true
     end)
```
apply this on this branch, an no-op stringify_data resolve the test.

ibhagwan commented:
> I walk the _G table with these two line, some _registry is always keeped in entry:

The registry is never cleared, even after `clear_protected` is called, it’s a ring buffer, function pointers (which hold ref to `opts`) only leave the registry after a full circle of the registry when a new entry overrides the same index.

> Still not sure why this line matter here:

This got me too, I looked at the generated command with debug=true, the only difference was the order of the callback function ids, for example in one the preview callback can be id 3 and in the other command (with _fzf_cli_args), it’s 1, otherwise the command is the same.

Another interesting fact is that I wasn’t able to reproduce this outside of testing, if I paste the same testing code into a lua file and `:source` it, I am only able to get the gc to be called with `previewer=false`, with “builtin” the gc is never called, with or  without `keymap.fzf`.

ibhagwan commented:
Interesting @phanen, what’s this for? Just to cycle the registry to be 1?
```lua
do  
  local _ = "execute-silent:" .. shell.stringify_data(function() end, {})
end
```

ibhagwan commented:
Interesting, changing the `do end` block to the below also makes the CI pass, as I suspected it's skipping a slot in the registry which makes the difference... wild stuff
```lua
  do
    local _ = shell.register_func(function() end)
  end
```

phanen commented:
I consider the parity of regostry MAX_LEN... but this seems irrelated.

Also need `defer_fn` replace the action... otherwise test still fail. 😢

ibhagwan commented:
> I consider the parity of regostry MAX_LEN... but this seems irrelated.
> 
> Also need `defer_fn` replace the action... otherwise test still fail. 😢

I found out min runs the test many times in the same child causing the registry to overflow and hence throw away the function ptr which causes the test to succeed, need another `shell.clear_registry()` which empties the registry, after which calling `collectgarbage` triggers the gc clear, it's working for me outside of the testing (in a simple lua source file)

Try sourcing this file:
```lua
require("fzf-lua")
FzfLua.setup({})
FzfLua.utils.setmetatable__gc = function(t, mt)
  local prox = newproxy(true)
  print("setting __gc proxy")
  getmetatable(prox).__gc = function()
    print("__gc called")
    mt.__gc(t)
  end
  t[prox] = true
  return setmetatable(t, mt)
end
FzfLua.files({
  previewer = "builtin",
  keymap = { fzf = { load = function() print("load called (keymap)") end } },
  -- actions = { load = { fn = function() print("load called (actions)") end, exec_silent = true, } },
})
```

Add this to `shell.lua`:
```lua
function M.clear_registry()
  _registry = {}
  _protected = {}
end
```

After hide call:
```lua
:lua FzfLua.shell.clear_registry()
:lua collectgarbage("collect")
```

And you'll see the message `__gc called`.

Still figuring out why it's not working in the test even after adding `clear_registry()`.

phanen commented:
yeah but better handle it in the api rather than the test if user shouldn't take care of `clear_registry()`.

ibhagwan commented:
> yeah but better handle it in the api rather than the test if user shouldn't take care of `clear_registry()`.

For sure, but this is just a test, technically we can replace `clear_protected` with `clear_registry`, unless we are resuming or unhiding I don't see any reason to keep any function pointers in the registry at all.

ibhagwan commented:
See https://github.com/ibhagwan/fzf-lua/pull/2152/commits/5c37803b1b42813733904deb01b488d6dceade28, yes `clear_registry` is better automated but I can't think where is the right place to put it, for example in testing since we use `_fzf_cli_args` in setup to trigger the load event calling `clear_registry` removes that pointer and messes up the test, hence:
https://github.com/ibhagwan/fzf-lua/blob/5c37803b1b42813733904deb01b488d6dceade28/tests/win_spec.lua#L60-L67

Now the only mystery left is why the api events test fails and why does it mess up all the other tests on Windows by adding `FZF_PROMPT/FZF_TOTAL_COUNT` to every single command output...

ibhagwan commented:
I’m thinking the best way forward would be to modify the function registry, ditch the “protected” idea, and make it an MRU store, once the store is full every new register would kick out the LRU function ptr.

As long as we have enough space in the store for one picker and its functions should never be an issue.

phanen commented:
I'm not sure I get it... if we only register temporary function, LRU will be naive as fifo:

Maybe you want to put entry when we call function? e.g.
```diff
diff --git a/lua/fzf-lua/shell.lua b/lua/fzf-lua/shell.lua
index dfd850a..34ef697 100644
--- a/lua/fzf-lua/shell.lua
+++ b/lua/fzf-lua/shell.lua
@@ -17,7 +17,7 @@ local _index = 0
 local _registry = {}
 local _protected = {}
 
-function M.register_func(fn)
+function M.lru_put(fn)
   repeat
     _index = _index % _MAX_LEN + 1
   until not _protected[_index]
@@ -59,7 +59,8 @@ end
 function M.pipe_wrap_fn(fn, fzf_field_index, debug)
   fzf_field_index = fzf_field_index or "{+}"
 
-  local receiving_function = function(pipe_path, ...)
+  local receiving_function
+  receiving_function = function(pipe_path, ...)
     local pipe = uv.new_pipe(false)
     local args = { ... }
     -- unescape double backslashes on windows
@@ -77,6 +78,7 @@ function M.pipe_wrap_fn(fn, fzf_field_index, debug)
           module.__INFO or {}, { selected = args[1][1] })
       end
     end)
+    M.lru_put(receiving_function)
     uv.pipe_connect(pipe, pipe_path, function(err)
       if err then
         error(string.format("pipe_connect(%s) failed with error: %s", pipe_path, err))
@@ -88,7 +90,7 @@ function M.pipe_wrap_fn(fn, fzf_field_index, debug)
     end)
   end
 
-  local id = M.register_func(receiving_function)
+  local id = M.lru_put(receiving_function)
 
   -- this is for windows WSL and AppImage users, their nvim path isn't just
   -- 'nvim', it can be something else
```

ibhagwan commented:
> Maybe you want to "access" when we call function? e.g.

Yes, that’s what I meant, when we “get” a function we bubble it to the top, so functions that are used will be cleared last.

> But I forgot why protected is needed?
> clear_protected only called when we start a new picker.

It’s not needed anymore, sometime in the past before reload i needed a way to tell actions that can be jded multiple times to never be cleared out of the ring buffer, for example git_status stage/unstage - but due to logic changes there is no such issue.

> set_protect only to avoid being wiped becuase call register_func within the picker's lifetime. If we use LRU, register_func new function will still wipe old one.

If we have enough room in the cache and we “access” on get then it won’t be deleted as it will be at the top of the cache.

> The problem is not to delete the non-frequent one but to avoid to delete unexpectedly.

Im trying to think when do we have a case of a non frequent function that should still be kept, maybe if the user has a custom action for a picker in `setup` and then doesn’t use the picker in a while?


> the problem is the timing of clear_registry , clear_protect if it can not be determine easily

Im not sure if it can be determined, since we have unhide/resume - we can never know if the user decided to resume a picker or start a new one.

Only if we store the function IDs of function pointers for each picker and clear only these ones? Sounds a bit over complicated.


> one more concern... lru feels non-determintic anther way is keep a pattern to never call shell.stringify outside of api

Maybe something in between? Allow it to be called but have some known limitations?


Bottom line, IMHO these are all edge cases that will only be encountered by the very sophisticated user customizing fzf-lua like what you did with fzf-lua-extra, in such case we can always guide the user to a better way of writing the code so I think LRU cache is the way to go here.

ibhagwan commented:
@phanen, https://github.com/ibhagwan/fzf-lua/pull/2152/commits/4ce85af7404e042e3ad7c6fa9a44b2108497c9e0.

Seems to work great but the GC test fails on Windows now, unsure why, shouldn't be a difference as all refs are clear, also unsure why this part?
https://github.com/ibhagwan/fzf-lua/blob/4ce85af7404e042e3ad7c6fa9a44b2108497c9e0/tests/win_spec.lua#L39-L46

Also, any ideas about how to fix the api events test on Windows and the weirdness that comes with it, if you look at the failed screenshots it somehow pollutes all other ui_spec tests with the output of the FZF_PROMPT / FZF_TOTAL COUNT env vars

ibhagwan commented:
https://github.com/ibhagwan/fzf-lua/pull/2152/commits/4ee6d64276e176a7317cab2b7480e1755783fac9

Alright, so I gave up on the Windows oddities with `%FZF_PROMPT%` messing up everything, instead on Windows we use `{fzf:prompt}`, it is deprecated but still working, at least we know the actions/execute-silent combo works:
https://github.com/ibhagwan/fzf-lua/blob/4ee6d64276e176a7317cab2b7480e1755783fac9/tests/api_spec.lua#L191-L202

I still don't know I'munable to use `%FZF_PROMPT%` on Windows, might be an fzf, fzf-lua or maybe just CI/mini.test bug? I can't test it as I no longer have access to Windows VM.

I restored the GC test to it's original code with a slight hack:
https://github.com/ibhagwan/fzf-lua/blob/4ee6d64276e176a7317cab2b7480e1755783fac9/tests/win_spec.lua#L39-L42

2 more tasks and this PR might be ready to merge:
- Get rid of `fn_pre_fzf`, `fn_post_fzf`, shouldn't be a big deal as it's used by only a few pickers
- Implement `reload_on_hide()` where it makes sense:
https://github.com/ibhagwan/fzf-lua/blob/4ee6d64276e176a7317cab2b7480e1755783fac9/lua/fzf-lua/profiles/hide.lua#L92-L97

phanen commented:
for gc i only know windows has a upstream bug it must manually kill the job 

not sure if %FZF_PROMPT% can be used... it's env var but i didn't see this in fzf's test case

ibhagwan commented:
Ok, with https://github.com/ibhagwan/fzf-lua/pull/2152/commits/787139925f39e1e17e561d68f0d1475b93c6b342 I feel this PR is almost done, could probably have a bit more polish here and there but most of what I wanted to achieve is complete.

I didn’t fully implement `_resume_reload` on all pickers where it matters and also a bit lazy on `blines` reloading indiscriminately (vs only when buffer changed) but the framework is there and it can easily be added at any time, any new picker that needs reloading we just add `_resume_reload = true` in its default options or for more complex scenarios `_resume_reload = function() … end`.

There are also obvious improvements to the `:FzfLua combine` picker which I snuck in this PR - all can be handled later.

We’ll give it a few days, see if we can think of some other changes or something will come up in testing.

ibhagwan commented:
@phanen, how do you feel about merging this?

ibhagwan commented:
With https://github.com/ibhagwan/fzf-lua/commit/34f7ef5a675fc88f3958012a16ff888f28b405da and some improvements to `combine` and `global` this PR should close #1879, #2058 and #2063.

phanen commented:
> how do you feel about merging this?

I think it's fine. The only break change for public api is `fzf_live`: https://github.com/search?q=fzf_live%20lang%3Alua%20NOT%20is%3Afork&type=code

ibhagwan commented:
> > how do you feel about merging this?
> 
> I think it's fine. The only break change for public api is `fzf_live`: https://github.com/search?q=fzf_live%20lang%3Alua%20NOT%20is%3Afork&type=code

Alright, I'll merge it tomorrow, see if you can think of any other issues.

Also, the new `global` picker has an issue with resume (that isn't an "unhide"), probably has more issues that I can't think about ATM but it's a new picker, I can fix post merge.

ibhagwan commented:
I’m not super happy with the hacky opts propagation in the new global picker, I’ll merge this anyway and work on this later, need to find a better solution that propagates consistent opts to both actions and convert actions (for reloading and hide profile).
